include/nlohmann/json.hpp

```diff
@@ -128,7 +129,7 @@ class basic_json // NOLINT(cppcoreguidelines-special-member-functions,hicpp-spec
 {
   private:
     template<detail::value_t> friend struct detail::external_constructor;
-    friend ::nlohmann::json_pointer<basic_json>;
+    friend ::nlohmann::json_pointer<StringType>;
```

Comment:
That declaration is correct and required for, e.g., `parent.erase(json_pointer::template array_index<basic_json_t>(last_path));`.

Reply:
It is required for `json_pointer<StringType>` but not for `json_pointer<basic_json>`?  Doesn't seem obvious why that's the case.  I guess if everything compiles, then it's not, just surprising.

Reply:
Because `json_pointer` is a typedef to `::nlohmann:json_pointer<StringType>`. **That** is the confusing typedef I was talking about in our `json_pointer` vs `::nlohmann:json_pointer` discussion.

Different file, different context, but this confusion you're experiencing now informed my stylistic choice then. ;-)

Reply:
Update: You were correct after all and there's a pretty big hole in the unit tests: the macro `JSON_TESTS_PRIVATE`.

We do need a friend declaration for the deprecated functions that use `nlohmann::json_pointer<basic_json>`.